src/main/cpp/src/ftos_converter.cuh

```diff
@@ -1424,4 +1475,15 @@ __device__ inline int format_float(double value, int digits, bool is_float, char
   }
 }
 
+//===== json_parser utility =====
+
+__device__ inline int double_normalization(double value, char* output)
+{
+  if (output == nullptr) {
```

Comment:
I haven't fully grasped the control flow, so this might be a silly observation:

We might be inadvertently checking `output == nullptr` multiple times, both from the call-site and here.  Since this is a `__device__` function, this might amount to a lot of calls.  

In a follow-up, there might be value in separating those call chains.  Computing the sizes can be done separately from actually writing the normalized value.  Just a thought for future consideration.

Reply:
Will fix in a follow-on